doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,192 @@
+.. _custom_datasources:
+
+==================
+Custom Datasources
+==================
+
+Ray Datasets supports multiple ways to :ref:`create a dataset <creating_datasets>`, 
+allowing you to easily ingest data of common formats from popular sources. However, if the 
+datasource you want to read from is not in the built-in list, don't worry, you can implement 
+a custom one for your use case. In this guide, we will walk you through how to build 
+your own custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.
+By the end of the guide, you will have a ``MongoDatasource`` that you can use to create dataset as follows:
+
+.. code-block:: python
+
+    # Read from custom MongoDB datasource to create a dataset.
+    ds = ray.data.read_datasource(
+        MongoDatasource(),
+        uri=MY_URI,
+        database=MY_DATABASE,
+        collection=MY_COLLECTION,
+        pipelines=MY_PIPELINES
+    )
+
+    # Write the dataset to custom MongoDB datasource.
+    ds.write_datasource(
+        MongoDatasource(), uri=MY_URI, database=MY_DATABASE, collection=MY_COLLECTION
+    )
+
+.. tip::
+
+    There are a few MongoDB concepts involved here. The `URI <https://www.mongodb.com/docs/manual/reference/connection-string/>`__ points to
+    a MongoDB instance, which hosts `Databases and Collections <https://www.mongodb.com/docs/manual/reference/connection-string/>`__. A collection
```

Comment:
this link should be https://www.mongodb.com/docs/manual/core/databases-and-collections/ .